importer/trickle/trickledag.go

```diff
@@ -56,3 +56,181 @@ func fillTrickleRec(db *h.DagBuilderHelper, node *h.UnixfsNode, depth int) error
 	}
 	return nil
 }
+
+// TrickleAppend appends the data in `db` to the dag, using the Trickledag format
+func TrickleAppend(base *dag.Node, db *h.DagBuilderHelper) (*dag.Node, error) {
+	// Convert to unixfs node for working with easily
+	ufsn, err := h.NewUnixfsNodeFromDag(base)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get depth of this 'tree'
+	n, j := trickleDepthInfo(base, db.Maxlinks())
+	if n == 0 {
+		// If direct blocks not filled...
+		err := db.FillNodeLayer(ufsn)
+		if err != nil {
+			return nil, err
+		}
+		n++
+
+		if db.Done() {
+			return ufsn.GetDagNode()
+		}
+	}
+
+	if len(base.Links) > db.Maxlinks() {
+		// Recursive step, grab last child
+		last := len(base.Links) - 1
+		lastChild, err := base.Links[last].GetNode(db.GetDagServ())
+		if err != nil {
+			return nil, err
+		}
+
+		// Fill out last child (may not be full tree)
+		nchild, err := trickleAppendRec(lastChild, db, n)
+		if err != nil {
+			return nil, err
+		}
+
+		// Update node link
+		ufsn.RemoveChild(last)
+		err = ufsn.AddChild(nchild, db)
+		if err != nil {
+			return nil, err
+		}
+
+		// Partially filled depth layer
+		if j != 0 {
+			for ; j < layerRepeat && !db.Done(); j++ {
+				next := h.NewUnixfsNode()
+				err := fillTrickleRec(db, next, n)
+				if err != nil {
+					return nil, err
+				}
+
+				err = ufsn.AddChild(next, db)
+				if err != nil {
+					return nil, err
+				}
+			}
+			n++
+		}
+
+	}
+
+	// Now, continue filling out tree like normal
+	for i := n; !db.Done(); i++ {
+		for j := 0; j < layerRepeat && !db.Done(); j++ {
+			next := h.NewUnixfsNode()
+			err := fillTrickleRec(db, next, i)
+			if err != nil {
+				return nil, err
+			}
+
+			err = ufsn.AddChild(next, db)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return ufsn.GetDagNode()
```

Comment:
maybe i'm just tired but i dont understand what the algorithm is doing. i roughly know what it's supposed to end up like, but maybe more doc would help very much.